{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import collections\n",
    "from collections import OrderedDict, Counter, defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from bokeh.plotting import figure, output_file, show \n",
    "import matplotlib.patches as patches    \n",
    "from matplotlib import colors\n",
    "\n",
    "\n",
    "import glob\n",
    "\n",
    "import subprocess\n",
    "from subprocess import call\n",
    "\n",
    "import re\n",
    "\n",
    "import pickle\n",
    "\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "import Bio.Data.CodonTable\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "import Bio.Data.CodonTable\n",
    "\n",
    "import joypy\n",
    "\n",
    "import pyBigWig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metadata G25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_tr</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>cds_seq</th>\n",
       "      <th>cds_start_codon</th>\n",
       "      <th>cds_stop_codon</th>\n",
       "      <th>cds_start_pos</th>\n",
       "      <th>cds_stop_pos</th>\n",
       "      <th>utr5_start</th>\n",
       "      <th>utr5_end</th>\n",
       "      <th>record_id</th>\n",
       "      <th>tr_id1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000335137.3</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5-001</td>\n",
       "      <td>ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATG</td>\n",
       "      <td>ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>0</td>\n",
       "      <td>918</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000335137.3|ENSG00000186092.4|OTTHUMG000...</td>\n",
       "      <td>ENST00000335137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>FO538757.2-201</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...</td>\n",
       "      <td>AAG</td>\n",
       "      <td>ATGAGTGACAGCATCAACTTCTCTCACAACCTAGGCCAGCTCCTGT...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>316</td>\n",
       "      <td>718</td>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>ENST00000624431.1|ENSG00000279928.1|-|-|FO5387...</td>\n",
       "      <td>ENST00000624431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id        gene         gene_tr  \\\n",
       "0  ENST00000335137.3       OR4F5       OR4F5-001   \n",
       "1  ENST00000624431.1  FO538757.2  FO538757.2-201   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...   \n",
       "1  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...   \n",
       "\n",
       "                                      5UTR_start_seq CDS_start  \\\n",
       "0                                                NaN       ATG   \n",
       "1  CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTA...       AAG   \n",
       "\n",
       "                                             cds_seq cds_start_codon  \\\n",
       "0  ATGGTGACTGAATTCATTTTTCTGGGTCTCTCTGATTCTCAGGAAC...             ATG   \n",
       "1  ATGAGTGACAGCATCAACTTCTCTCACAACCTAGGCCAGCTCCTGT...             ATG   \n",
       "\n",
       "  cds_stop_codon  cds_start_pos  cds_stop_pos  utr5_start  utr5_end  \\\n",
       "0            TAG              0           918           0         0   \n",
       "1            TAG            316           718           0       316   \n",
       "\n",
       "                                           record_id           tr_id1  \n",
       "0  ENST00000335137.3|ENSG00000186092.4|OTTHUMG000...  ENST00000335137  \n",
       "1  ENST00000624431.1|ENSG00000279928.1|-|-|FO5387...  ENST00000624431  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/home/alla/make_PhD_great_again/NTE/VERSION_since_May2021/'\n",
    "\n",
    "# gencode 25\n",
    "metadata_pc_g25 = pd.read_csv(path+'tmp_res/metadata_pc_g25.txt', sep='\\t')\n",
    "metadata_pc_g25['tr_id1'] = [x.split('.')[0] for x in metadata_pc_g25['tr_id'].tolist()] \n",
    "\n",
    "metadata_pc_g25[0:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### get only complete CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "94359 78873 60109\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alla/.local/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "tmp = metadata_pc_g25[(metadata_pc_g25['utr5_end'].notna())]\n",
    "tmp2 = tmp[tmp['utr5_end'] > 0]\n",
    "\n",
    "tmp2['tr_len'] = tmp2['transcript_seq'].str.len()\n",
    "tmp3 = tmp2[tmp2['tr_len'] > tmp2['cds_stop_pos']]\n",
    "\n",
    "print (metadata_pc_g25.shape[0], tmp2.shape[0], tmp3.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = dict(tuple(tmp3[['tr_id', 'gene', 'cds_start_codon', 'utr5_end']].groupby(by='gene')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AL008723.1 ENST00000623133.1 CTG\n",
      "ARAP1 ENST00000455638.6 ATA\n",
      "ATP13A3 ENST00000619199.4 ATT\n",
      "BAG1 ENST00000472232.7 CTG\n",
      "C16orf82 ENST00000418886.1 CTG\n",
      "C1orf234 ENST00000622840.1 AGG\n",
      "CACNG8 ENST00000270458.2 CTG\n",
      "CASK ENST00000378179.7 CTG\n",
      "DDX17 ENST00000396821.7 CTG\n",
      "DDX17 ENST00000403230.1 CTG\n",
      "EBF1 ENST00000622875.4 AGG\n",
      "EIF4G2 ENST00000526148.5 GTG\n",
      "EIF4G2 ENST00000525681.5 GTG\n",
      "EIF4G2 ENST00000339995.9 GTG\n",
      "EIF4G2 ENST00000396525.6 GTG\n",
      "EIF4G2 ENST00000531647.5 GTG\n",
      "EIF4G2 ENST00000528562.5 GTG\n",
      "FGF2 ENST00000614010.4 CTG\n",
      "FGF2 ENST00000264498.7 CTG\n",
      "FNDC5 ENST00000373471.7 ATA\n",
      "GTF3A ENST00000381140.8 CTG\n",
      "GTF3A ENST00000435781.5 CTG\n",
      "GTF3A ENST00000419181.5 CTG\n",
      "GTF3A ENST00000438571.5 CTG\n",
      "HCK ENST00000534862.5 CTG\n",
      "HCK ENST00000486475.5 CTG\n",
      "HCK ENST00000262651.3 CTG\n",
      "HCK ENST00000375852.2 CTG\n",
      "HMHB1 ENST00000289448.2 CTG\n",
      "KCNIP1 ENST00000390656.8 AAG\n",
      "MDFIC ENST00000257724.7 GTG\n",
      "MDFIC ENST00000614186.4 GTG\n",
      "MDGA2 ENST00000426342.5 TTG\n",
      "MRVI1 ENST00000527509.6 CTG\n",
      "MYC ENST00000613283.1 CTG\n",
      "MYC ENST00000621592.5 CTG\n",
      "MYC ENST00000524013.1 CTG\n",
      "NELFB ENST00000343053.4 CTG\n",
      "NPTXR ENST00000333039.2 CTG\n",
      "NPW ENST00000329610.4 CTG\n",
      "NR1I2 ENST00000393716.6 CTG\n",
      "NR1I2 ENST00000466380.5 CTG\n",
      "OAZ3 ENST00000479764.6 CTG\n",
      "OAZ3 ENST00000400999.6 CTG\n",
      "PEG10 ENST00000615790.4 CTG\n",
      "PEG10 ENST00000612941.1 CTG\n",
      "PIGX ENST00000392391.7 CTG\n",
      "POU5F1 ENST00000620031.4 CTG\n",
      "POU5F1 ENST00000606567.5 CTG\n",
      "PRAMEF18 ENST00000624297.1 ATC\n",
      "PRPS1L1 ENST00000506618.3 ACG\n",
      "PRR16 ENST00000446965.2 GTG\n",
      "R3HCC1 ENST00000522012.5 CTG\n",
      "R3HCC1 ENST00000625275.2 CTG\n",
      "R3HCC1 ENST00000265806.10 CTG\n",
      "RNF187 ENST00000305943.7 CTG\n",
      "SCNN1D ENST00000616713.4 ATC\n",
      "SLC22A17 ENST00000397260.7 CTG\n",
      "SLC22A17 ENST00000474057.5 CTG\n",
      "SLC22A17 ENST00000636431.1 CTG\n",
      "SORCS1 ENST00000612154.4 ATA\n",
      "SORCS1 ENST00000622431.4 ATA\n",
      "SP3 ENST00000418194.6 ATA\n",
      "STIM2 ENST00000467087.5 TTG\n",
      "STIM2 ENST00000467011.5 TTG\n",
      "STIM2 ENST00000465503.5 TTG\n",
      "STK11IP ENST00000295641.14 AGG\n",
      "SWI5 ENST00000608796.5 CTG\n",
      "SWI5 ENST00000419867.6 CTG\n",
      "SWI5 ENST00000418976.1 CTG\n",
      "TEAD1 ENST00000527636.5 ATT\n",
      "TEAD1 ENST00000527575.5 ATT\n",
      "TEAD3 ENST00000338863.11 ATA\n",
      "TEAD4 ENST00000358409.6 TTG\n",
      "TEAD4 ENST00000359864.6 TTG\n",
      "TEAD4 ENST00000443986.2 TTG\n",
      "TGM5 ENST00000622115.1 CTG\n",
      "TRPV6 ENST00000359396.7 ACG\n",
      "TXNRD3 ENST00000524230.6 CTG\n",
      "TXNRD3 ENST00000523403.2 CTG\n",
      "VEGFA ENST00000611736.4 CTG\n",
      "VEGFA ENST00000621747.4 CTG\n",
      "VEGFA ENST00000615393.4 CTG\n",
      "VEGFA ENST00000617771.4 CTG\n",
      "VEGFA ENST00000372067.7 CTG\n",
      "WT1 ENST00000379077.7 CTG\n",
      "WT1 ENST00000332351.7 CTG\n",
      "WT1 ENST00000448076.7 CTG\n",
      "ZNF804A ENST00000613975.1 AGG\n"
     ]
    }
   ],
   "source": [
    "for gene, df in d.items():\n",
    "    starts = set()\n",
    "    for el in df.to_numpy():\n",
    "        tr_id = el[0]\n",
    "        cds_start_codon = el[2]\n",
    "        if cds_start_codon in ['CTG', 'TTG', 'GTG', 'ACG', 'AGG', 'AAG', 'ATT', 'ATA', 'ATC']:\n",
    "            print (gene, tr_id, cds_start_codon)\n",
    "        starts.add(cds_start_codon)\n",
    "    \n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = {'EIF4G2': [493, 1111, 'ENST00000339995'],  # nonAUG, dAUG1, tr_id \n",
    "     'BAG1': [88, 301, 'ENST00000472232'], \n",
    "     'R3HCC1': [269, 830, 'ENST00000265806'], \n",
    "     'GTF3A': [125, 900, 'ENST00000381140'], \n",
    "     'DDX17': [76, 313, 'ENST00000403230'], \n",
    "     'RNF187': [429, 756, 'ENST00000305943'],\n",
    "     'HCK': [229, 292, 'ENST00000375862'], \n",
    "     'NELFB': [194, 338, 'ENST00000343053'],\n",
    "     \n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "AL008723.1 ENST00000623133.1 CTG - not in G41 anymore \n",
    "\n",
    "ARAP1 ENST00000455638.6 ATA - removed \n",
    "\n",
    "ATP13A3 ENST00000619199.4 ATT - not in G41 anymore \n",
    "\n",
    "BAG1 ENST00000472232.7 CTG - OK\n",
    "\n",
    "C16orf82 ENST00000418886.1 CTG - testis-specific, has goos MS spectra  \n",
    "\n",
    "C1orf234 ENST00000622840.1 AGG - not in G41 anymore\n",
    "\n",
    "CACNG8 ENST00000270458.2 CTG\n",
    "\n",
    "CASK ENST00000378179.7 CTG\n",
    "\n",
    "DDX17 ENST00000396821.7 CTG  OK\n",
    "DDX17 ENST00000403230.1 CTG  OK \n",
    "\n",
    "EBF1 ENST00000622875.4 AGG - will be removed \n",
    "\n",
    "EIF4G2 ENST00000526148.5 GTG\n",
    "EIF4G2 ENST00000525681.5 GTG\n",
    "EIF4G2 ENST00000339995.9 GTG\n",
    "EIF4G2 ENST00000396525.6 GTG\n",
    "EIF4G2 ENST00000531647.5 GTG\n",
    "EIF4G2 ENST00000528562.5 GTG\n",
    "\n",
    "FGF2 ENST00000614010.4 CTG\n",
    "FGF2 ENST00000264498.7 CTG\n",
    "\n",
    "FNDC5 ENST00000373471.7 ATA - ? both ATG and ATA. no support for exon with ATA. Upstream ATG is annotated in human and mouse  \n",
    "\n",
    "GTF3A ENST00000381140.8 CTG\n",
    "GTF3A ENST00000435781.5 CTG\n",
    "GTF3A ENST00000419181.5 CTG\n",
    "GTF3A ENST00000438571.5 CTG\n",
    "\n",
    "HCK ENST00000534862.5 CTG - a lot of transcripts, including in-frame cdsATGs, CTG seems genuine \n",
    "HCK ENST00000486475.5 CTG\n",
    "HCK ENST00000262651.3 CTG\n",
    "HCK ENST00000375852.2 CTG\n",
    "\n",
    "HMHB1 ENST00000289448.2 CTG - ? not conserved , rna-seq is in intron etc. weird gene in MANE, abberant one, kept \n",
    "\n",
    "KCNIP1 ENST00000390656.8 AAG - rna-seq/ribo-seq in intron... there is tr_id with dATG as a start it's removed now\n",
    "\n",
    "MDFIC ENST00000257724.7 GTG gtg is not conserved ribo-seq is not convincing it's gone in g41\n",
    "MDFIC ENST00000614186.4 GTG\n",
    "\n",
    "---------------\n",
    "\n",
    "MDGA2 ENST00000426342.5 TTG more like internal one, no ribo-seq support \n",
    "\n",
    "MRVI1 ENST00000527509.6 CTG, internal CTG, no ribo-seq \n",
    "\n",
    "MYC ENST00000613283.1 CTG\n",
    "MYC ENST00000621592.5 CTG\n",
    "MYC ENST00000524013.1 CTG\n",
    "\n",
    "NELFB ENST00000343053.4 CTG exslucive nonAUG/predominant? sink ORF \n",
    "\n",
    "NPTXR ENST00000333039.2 CTG exslucive nonAUG/predominant? dAUG is not even annotated \n",
    "\n",
    "NPW ENST00000329610.4 CTG - predominant CTG. but there is another weird transcript with upstream exon no support \n",
    "\n",
    "NR1I2 ENST00000393716.6 CTG  - no ribo-seq for this CTG, CTG is ~ok in mammals as well as dAUG. why was it annotated in a first place  \n",
    "NR1I2 ENST00000466380.5 CTG\n",
    "\n",
    "OAZ3 ENST00000479764.6 CTG\n",
    "OAZ3 ENST00000400999.6 CTG\n",
    "\n",
    "PEG10 ENST00000615790.4 CTG\n",
    "PEG10 ENST00000612941.1 CTG\n",
    "\n",
    "PIGX ENST00000392391.7 CTG exclusive no sink there is dAUG isoform wit shorter 5'leader \n",
    "\n",
    "POU5F1 ENST00000620031.4 CTG no ribo-seq more like internal one \n",
    "POU5F1 ENST00000606567.5 CTG\n",
    "\n",
    "PRAMEF18 ENST00000624297.1 ATC removed in G41 \n",
    " \n",
    "PRPS1L1 ENST00000506618.3 ACG - one exon gene with mappability issue. in trips looks good with amb reads like predominantly; it's ACG in human but ATG in other species. pc signal is fine \n",
    "\n",
    "PRR16 ENST00000446965.2 GTG, thre is dATG. could be exclusive, sink ORF \n",
    "\n",
    "R3HCC1 ENST00000522012.5 CTG\n",
    "R3HCC1 ENST00000625275.2 CTG\n",
    "R3HCC1 ENST00000265806.10 CTG\n",
    "\n",
    "RNF187 ENST00000305943.7 CTG excl ctg \n",
    "\n",
    "SCNN1D ENST00000616713.4 ATC not a lot of ribo-seq, is not in G41 anymore \n",
    "\n",
    "SLC22A17 ENST00000397260.7 CTG exlusive CTG? G41 has only CTG. G25 and refseq has dATG \n",
    "SLC22A17 ENST00000474057.5 CTG\n",
    "SLC22A17 ENST00000636431.1 CTG\n",
    "\n",
    "SORCS1 ENST00000612154.4 ATA  ATA right in front of ATG. there is upstream ATG. no ribo-seq.  \n",
    "SORCS1 ENST00000622431.4 ATA\n",
    "\n",
    "SP3 ENST00000418194.6 ATA could be predominant in this articular trascrniot , there are CUGCUG \n",
    "\n",
    "STIM2 ENST00000467087.5 TTG ok but there was upstream ATG \n",
    "STIM2 ENST00000467011.5 TTG\n",
    "STIM2 ENST00000465503.5 TTG\n",
    "\n",
    "STK11IP ENST00000295641.14 AGG it's not in G41. but in trips-viz it looks suprisingly okay as a start  \n",
    "\n",
    "SWI5 ENST00000608796.5 CTG looks like exclusive nonAUG\n",
    "SWI5 ENST00000419867.6 CTG\n",
    "SWI5 ENST00000418976.1 CTG\n",
    "\n",
    "TEAD1 ENST00000527636.5 ATT\n",
    "TEAD1 ENST00000527575.5 ATT\n",
    "\n",
    "TEAD3 ENST00000338863.11 ATA\n",
    "\n",
    "TEAD4 ENST00000358409.6 TTG\n",
    "TEAD4 ENST00000359864.6 TTG\n",
    "TEAD4 ENST00000443986.2 TTG\n",
    "\n",
    "TGM5 ENST00000622115.1 CTG weird short exon; does not exist in G41 \n",
    "\n",
    "TRPV6 ENST00000359396.7 ACG\n",
    "\n",
    "TXNRD3 ENST00000524230.6 CTG exclusive \n",
    "TXNRD3 ENST00000523403.2 CTG\n",
    "\n",
    "VEGFA ENST00000611736.4 CTG\n",
    "VEGFA ENST00000621747.4 CTG\n",
    "VEGFA ENST00000615393.4 CTG\n",
    "VEGFA ENST00000617771.4 CTG\n",
    "VEGFA ENST00000372067.7 CTG\n",
    "\n",
    "WT1 ENST00000379077.7 CTG there are CTG and ATG peaks \n",
    "WT1 ENST00000332351.7 CTG\n",
    "WT1 ENST00000448076.7 CTG\n",
    "\n",
    "ZNF804A ENST00000613975.1 AGG, a lot of rna-seq data in introns; no such isoform in G41"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Transcripts/genes with predominant/exclusive nonAUG from annotation & ribo-seq. GENCODE25 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = [\n",
    "    ['EIF4G2', 493, 1111, 'ENST00000339995', 'sink_ORF'],  # nonAUG, dAUG1, tr_id \n",
    "     ['BAG1', 88, 301, 'ENST00000472232', 'sink_ORF'], \n",
    "     ['R3HCC1', 269, 830, 'ENST00000265806', 'sink_ORF'], \n",
    "     ['GTF3A', 125, 900, 'ENST00000381140', 'sink_ORF'], \n",
    "     ['DDX17', 76, 313, 'ENST00000403230', 'sink_ORF'], \n",
    "     ['RNF187', 429, 756, 'ENST00000305943', 'sink_ORF'],\n",
    "     ['HCK', 229, 292, 'ENST00000375862', 'sink_ORF'], \n",
    "     ['NELFB', 194, 338, 'ENST00000343053', 'sink_ORF'], \n",
    "     ['NPTXR', 125, 158, 'ENST00000333039', 'sink_ORF'], \n",
    "     ['NPW', 383, 539, 'ENST00000329610', 'sink_ORF'], \n",
    "     ['PIGX', 176, 299, 'ENST00000392391', 'sink_ORF'],\n",
    "     ['PRPS1L1', 82, 282,'ENST00000506618', 'sink_ORF'],\n",
    "     ['PRR16', 181, 232, 'ENST00000446965', 'sink_ORF'],\n",
    "     ['SLC22A17', 120, 453, 'ENST00000397260', 'sink_ORF'], \n",
    "     ['SP3', 385, 1036, 'ENST00000418194', 'sink_ORF'],\n",
    "     ['SWI5', 120, 438, 'ENST00000608796', 'sink_ORF'],\n",
    "     ['TXNRD3', 166, 712, 'ENST00000524230', 'sink_ORF'],\n",
    "     ['CCDC8', ...]   # annotated in GENCODE41, we take correcponsing GENCODE25 transcript and move CDS annotation \n",
    "\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>nonAUG_start</th>\n",
       "      <th>AUG1_start</th>\n",
       "      <th>tr_id1</th>\n",
       "      <th>mechanism</th>\n",
       "      <th>tr_id</th>\n",
       "      <th>dist_between_AUG_nonAUG_codons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NPTXR</td>\n",
       "      <td>125</td>\n",
       "      <td>158</td>\n",
       "      <td>ENST00000333039</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000333039.2</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PRR16</td>\n",
       "      <td>181</td>\n",
       "      <td>232</td>\n",
       "      <td>ENST00000446965</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000446965.2</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HCK</td>\n",
       "      <td>229</td>\n",
       "      <td>292</td>\n",
       "      <td>ENST00000375862</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000375862.6</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PIGX</td>\n",
       "      <td>176</td>\n",
       "      <td>299</td>\n",
       "      <td>ENST00000392391</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000392391.7</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NELFB</td>\n",
       "      <td>194</td>\n",
       "      <td>338</td>\n",
       "      <td>ENST00000343053</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000343053.4</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NPW</td>\n",
       "      <td>383</td>\n",
       "      <td>539</td>\n",
       "      <td>ENST00000329610</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000329610.4</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PRPS1L1</td>\n",
       "      <td>82</td>\n",
       "      <td>282</td>\n",
       "      <td>ENST00000506618</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000506618.3</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BAG1</td>\n",
       "      <td>88</td>\n",
       "      <td>301</td>\n",
       "      <td>ENST00000472232</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000472232.7</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DDX17</td>\n",
       "      <td>76</td>\n",
       "      <td>313</td>\n",
       "      <td>ENST00000403230</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000403230.1</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SWI5</td>\n",
       "      <td>120</td>\n",
       "      <td>438</td>\n",
       "      <td>ENST00000608796</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000608796.5</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RNF187</td>\n",
       "      <td>429</td>\n",
       "      <td>756</td>\n",
       "      <td>ENST00000305943</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000305943.7</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SLC22A17</td>\n",
       "      <td>120</td>\n",
       "      <td>453</td>\n",
       "      <td>ENST00000397260</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000397260.7</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TXNRD3</td>\n",
       "      <td>166</td>\n",
       "      <td>712</td>\n",
       "      <td>ENST00000524230</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000524230.6</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R3HCC1</td>\n",
       "      <td>269</td>\n",
       "      <td>830</td>\n",
       "      <td>ENST00000265806</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000265806.10</td>\n",
       "      <td>187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>493</td>\n",
       "      <td>1111</td>\n",
       "      <td>ENST00000339995</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000339995.9</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SP3</td>\n",
       "      <td>385</td>\n",
       "      <td>1036</td>\n",
       "      <td>ENST00000418194</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000418194.6</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTF3A</td>\n",
       "      <td>125</td>\n",
       "      <td>900</td>\n",
       "      <td>ENST00000381140</td>\n",
       "      <td>sink_ORF</td>\n",
       "      <td>ENST00000381140.8</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene  nonAUG_start  AUG1_start           tr_id1 mechanism  \\\n",
       "8      NPTXR           125         158  ENST00000333039  sink_ORF   \n",
       "12     PRR16           181         232  ENST00000446965  sink_ORF   \n",
       "6        HCK           229         292  ENST00000375862  sink_ORF   \n",
       "10      PIGX           176         299  ENST00000392391  sink_ORF   \n",
       "7      NELFB           194         338  ENST00000343053  sink_ORF   \n",
       "9        NPW           383         539  ENST00000329610  sink_ORF   \n",
       "11   PRPS1L1            82         282  ENST00000506618  sink_ORF   \n",
       "1       BAG1            88         301  ENST00000472232  sink_ORF   \n",
       "4      DDX17            76         313  ENST00000403230  sink_ORF   \n",
       "15      SWI5           120         438  ENST00000608796  sink_ORF   \n",
       "5     RNF187           429         756  ENST00000305943  sink_ORF   \n",
       "13  SLC22A17           120         453  ENST00000397260  sink_ORF   \n",
       "16    TXNRD3           166         712  ENST00000524230  sink_ORF   \n",
       "2     R3HCC1           269         830  ENST00000265806  sink_ORF   \n",
       "0     EIF4G2           493        1111  ENST00000339995  sink_ORF   \n",
       "14       SP3           385        1036  ENST00000418194  sink_ORF   \n",
       "3      GTF3A           125         900  ENST00000381140  sink_ORF   \n",
       "\n",
       "                 tr_id  dist_between_AUG_nonAUG_codons  \n",
       "8    ENST00000333039.2                              11  \n",
       "12   ENST00000446965.2                              17  \n",
       "6    ENST00000375862.6                              21  \n",
       "10   ENST00000392391.7                              41  \n",
       "7    ENST00000343053.4                              48  \n",
       "9    ENST00000329610.4                              52  \n",
       "11   ENST00000506618.3                              66  \n",
       "1    ENST00000472232.7                              71  \n",
       "4    ENST00000403230.1                              79  \n",
       "15   ENST00000608796.5                             106  \n",
       "5    ENST00000305943.7                             109  \n",
       "13   ENST00000397260.7                             111  \n",
       "16   ENST00000524230.6                             182  \n",
       "2   ENST00000265806.10                             187  \n",
       "0    ENST00000339995.9                             206  \n",
       "14   ENST00000418194.6                             217  \n",
       "3    ENST00000381140.8                             258  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(d, columns = ['gene', 'nonAUG_start', 'AUG1_start', 'tr_id1', 'mechanism']).merge(metadata_pc_g25[['tr_id', 'tr_id1']], on='tr_id1', how='inner')\n",
    "\n",
    "df['dist_between_AUG_nonAUG_codons'] = (df['AUG1_start'] - df['nonAUG_start']) // 3\n",
    "\n",
    "df.sort_values(by='dist_between_AUG_nonAUG_codons')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "create profiles for all of them \n",
    "alignments: avoidaance of in-frame AUG? \n",
    "sink ORFs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metadata G41"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extrac_metadata(filepath, pc=True):\n",
    "    if pc==True:\n",
    "        troubles = []\n",
    "        li = []\n",
    "        with open(filepath, \"r\") as handle:\n",
    "            for record in SeqIO.parse(handle, \"fasta\"):\n",
    "                record_id = str(record.id)\n",
    "                tr_id = record_id.split('|')[0]\n",
    "                seq = str(record.seq)\n",
    "                \n",
    "                gene = record_id.split('|')[5]\n",
    "                gene_tr = record_id.split('|')[4]\n",
    "                \n",
    "                # CDS extraction\n",
    "                m_cds = re.finditer(r'(CDS:)[0-9]*-[0-9]*', record_id)\n",
    "                cds = [x.group(0) for x in m_cds][0] #UTR3:955-1054\n",
    "                cds_start = int(cds.split(':')[1].split('-')[0]) - 1\n",
    "                cds_end = int(cds.split(':')[1].split('-')[1]) \n",
    "                cds_seq = seq[cds_start:cds_end]     \n",
    "                cds_start_codon = cds_seq[0:3]\n",
    "                cds_stop_codon = cds_seq[-3:]\n",
    "        \n",
    "                \n",
    "                # UTR5 extraction\n",
    "                # some transcript do not have UTR5 annotated \n",
    "                # but we still want to save such trascripts\n",
    "                m_utr5 = re.finditer(r'(UTR5:)[0-9]*-[0-9]*', record_id)\n",
    "                t = [x.group(0) for x in m_utr5]\n",
    "                \n",
    "                \n",
    "                if t:\n",
    "                    utr5 = t[0] #UTR5:1-1054\n",
    "                    utr5_start = int(utr5.split(':')[1].split('-')[0]) - 1\n",
    "                    utr5_end = int(utr5.split(':')[1].split('-')[1]) \n",
    "        \n",
    "                    if utr5_start == utr5_end: #1=1\n",
    "                        print ('utr5_start == utr5_end', record_id)\n",
    "                        utr5_start_seq = seq[utr5_start-1:(utr5_end)]\n",
    "                    else: \n",
    "                        utr5_start_seq = seq[utr5_start:(utr5_end)]\n",
    "                        \n",
    "                    li.append([tr_id, gene, gene_tr, seq, utr5_start_seq, \n",
    "                   utr5_start_seq[-3:], cds_seq,\n",
    "                  cds_start_codon, cds_stop_codon, cds_start, cds_end,\n",
    "                  utr5_start, utr5_end, record_id])\n",
    "                    \n",
    "                    \n",
    "                else:\n",
    "                    utr5_start = 0\n",
    "                    utr5_end = 0\n",
    "                    utr5_start_seq = None\n",
    "                    cds_start = 0\n",
    "                    \n",
    "                    li.append([tr_id, gene, gene_tr, seq, utr5_start_seq, \n",
    "                   cds_seq[:3], cds_seq,\n",
    "                  cds_start_codon, cds_stop_codon, cds_start, cds_end,\n",
    "                  utr5_start, utr5_end, record_id])\n",
    "        \n",
    "        metadata_pc_g25 = pd.DataFrame(li, columns = ['tr_id', 'gene' ,'gene_tr', \n",
    "                                 'transcript_seq', '5UTR_start_seq', \n",
    "                                 'CDS_start', 'cds_seq', 'cds_start_codon', \n",
    "                                 'cds_stop_codon', 'cds_start_pos',  'cds_stop_pos',\n",
    "                                 'utr5_start', 'utr5_end', 'record_id'])\n",
    "        return (metadata_pc_g25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# trs with 5UTR: 93459 ; # all trs 110224\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_tr</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>5UTR_start_seq</th>\n",
       "      <th>CDS_start</th>\n",
       "      <th>cds_seq</th>\n",
       "      <th>cds_start_codon</th>\n",
       "      <th>cds_stop_codon</th>\n",
       "      <th>cds_start_pos</th>\n",
       "      <th>cds_stop_pos</th>\n",
       "      <th>utr5_start</th>\n",
       "      <th>utr5_end</th>\n",
       "      <th>record_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000641515.2</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>OR4F5-201</td>\n",
       "      <td>CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...</td>\n",
       "      <td>CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...</td>\n",
       "      <td>GTT</td>\n",
       "      <td>ATGAAGAAGGTAACTGCAGAGGCTATTTCCTGGAATGAATCAACGA...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAG</td>\n",
       "      <td>60</td>\n",
       "      <td>1041</td>\n",
       "      <td>0</td>\n",
       "      <td>60</td>\n",
       "      <td>ENST00000641515.2|ENSG00000186092.7|OTTHUMG000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000426406.4</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>OR4F29-201</td>\n",
       "      <td>ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...</td>\n",
       "      <td>None</td>\n",
       "      <td>ATG</td>\n",
       "      <td>ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>TAA</td>\n",
       "      <td>0</td>\n",
       "      <td>939</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENST00000426406.4|ENSG00000284733.2|OTTHUMG000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id    gene     gene_tr  \\\n",
       "0  ENST00000641515.2   OR4F5   OR4F5-201   \n",
       "1  ENST00000426406.4  OR4F29  OR4F29-201   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...   \n",
       "1  ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...   \n",
       "\n",
       "                                      5UTR_start_seq CDS_start  \\\n",
       "0  CCCAGATCTCTTCAGTTTTTATGCCTCATTCTGTGAAAATTGCTGT...       GTT   \n",
       "1                                               None       ATG   \n",
       "\n",
       "                                             cds_seq cds_start_codon  \\\n",
       "0  ATGAAGAAGGTAACTGCAGAGGCTATTTCCTGGAATGAATCAACGA...             ATG   \n",
       "1  ATGGATGGAGAGAATCACTCAGTGGTATCTGAGTTTTTGTTTCTGG...             ATG   \n",
       "\n",
       "  cds_stop_codon  cds_start_pos  cds_stop_pos  utr5_start  utr5_end  \\\n",
       "0            TAG             60          1041           0        60   \n",
       "1            TAA              0           939           0         0   \n",
       "\n",
       "                                           record_id  \n",
       "0  ENST00000641515.2|ENSG00000186092.7|OTTHUMG000...  \n",
       "1  ENST00000426406.4|ENSG00000284733.2|OTTHUMG000...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_pc_g41 = extrac_metadata(filepath=\"data/gencode.v41.pc_transcripts.fa\", pc=True)\n",
    "print ('# trs with 5UTR:', metadata_pc_g41[metadata_pc_g41['5UTR_start_seq'].notna()].shape[0], '; # all trs', metadata_pc_g41.shape[0])\n",
    "metadata_pc_g41.to_csv('tmp_res/metadata_pc_g41.txt', sep='\\t', index=False)\n",
    "\n",
    "metadata_pc_g41[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alla/.local/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  after removing the cwd from sys.path.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ARAP1 ENST00000455638.6 ATA\n",
      "BAG1 ENST00000634734.3 CTG\n",
      "C16orf82 ENST00000505035.3 CTG\n",
      "CACNG8 ENST00000270458.4 CTG\n",
      "CCDC8 ENST00000697726.1 CTG\n",
      "DDX17 ENST00000403230.3 CTG\n",
      "DDX17 ENST00000396821.8 CTG\n",
      "EBF1 ENST00000622875.4 AGG\n",
      "EFCAB14 ENST00000672422.2 CTG\n",
      "EIF4G2 ENST00000526148.6 GTG\n",
      "EIF4G2 ENST00000525681.6 GTG\n",
      "EIF4G2 ENST00000339995.11 GTG\n",
      "EIF4G2 ENST00000396525.7 GTG\n",
      "EIF4G2 ENST00000531647.6 GTG\n",
      "EIF4G2 ENST00000528562.6 GTG\n",
      "FGF2 ENST00000264498.8 CTG\n",
      "FNDC5 ENST00000373471.9 ATA\n",
      "FNDC5 ENST00000649537.2 ATA\n",
      "GTF3A ENST00000381140.10 CTG\n",
      "GTF3A ENST00000435781.5 CTG\n",
      "GTF3A ENST00000419181.5 CTG\n",
      "GTF3A ENST00000438571.5 CTG\n",
      "HCK ENST00000486475.6 CTG\n",
      "HCK ENST00000262651.4 CTG\n",
      "HCK ENST00000375852.5 CTG\n",
      "HMHB1 ENST00000289448.4 CTG\n",
      "IRAG1 ENST00000527509.7 CTG\n",
      "KCNIP1 ENST00000390656.8 AAG\n",
      "KCTD11 ENST00000333751.8 ATT\n",
      "MDGA2 ENST00000426342.7 TTG\n",
      "MORN2 ENST00000644631.4 CTG\n",
      "MRTFA ENST00000651595.2 GTG\n",
      "MRTFA ENST00000355630.10 GTG\n",
      "MRTFA ENST00000402042.7 GTG\n",
      "MYC ENST00000524013.2 CTG\n",
      "MYC ENST00000621592.8 CTG\n",
      "NELFB ENST00000343053.6 CTG\n",
      "NFKBID ENST00000641389.2 CTG\n",
      "NPTXR ENST00000333039.4 CTG\n",
      "NPW ENST00000566435.4 CTG\n",
      "NR1I2 ENST00000466380.6 CTG\n",
      "NR1I2 ENST00000393716.8 CTG\n",
      "OAZ3 ENST00000479764.7 CTG\n",
      "OAZ3 ENST00000400999.7 CTG\n",
      "OR8G3P ENST00000641572.1 GTG\n",
      "OR8G3P ENST00000641848.1 GTG\n",
      "PANK2 ENST00000643504.2 CTG\n",
      "PANK2 ENST00000610179.7 CTG\n",
      "PANK2 ENST00000336066.8 CTG\n",
      "PEG10 ENST00000615790.5 CTG\n",
      "PEG10 ENST00000612941.2 CTG\n",
      "PIGX ENST00000392391.9 CTG\n",
      "POU5F1 ENST00000606567.6 CTG\n",
      "PRPS1L1 ENST00000506618.5 ACG\n",
      "PRR16 ENST00000446965.2 GTG\n",
      "PTEN ENST00000693560.1 CTG\n",
      "R3HCC1 ENST00000522012.6 CTG\n",
      "R3HCC1 ENST00000625275.3 CTG\n",
      "R3HCC1 ENST00000265806.12 CTG\n",
      "RNF187 ENST00000305943.9 CTG\n",
      "RPTOR ENST00000697423.1 CTG\n",
      "SFPQ ENST00000696553.1 GTG\n",
      "SLC22A17 ENST00000397260.7 CTG\n",
      "SLC22A17 ENST00000397267.6 CTG\n",
      "SLC22A17 ENST00000474057.6 CTG\n",
      "SLC22A17 ENST00000354772.10 CTG\n",
      "SLC22A17 ENST00000636431.2 CTG\n",
      "SLC6A1 ENST00000698198.1 CTG\n",
      "SORCS1 ENST00000612154.4 ATA\n",
      "SORCS1 ENST00000622431.4 ATA\n",
      "SP3 ENST00000418194.7 ATA\n",
      "SP3 ENST00000652005.2 ATA\n",
      "STIM2 ENST00000467011.6 TTG\n",
      "STIM2 ENST00000465503.6 TTG\n",
      "STIM2 ENST00000467087.7 TTG\n",
      "SWI5 ENST00000608796.6 CTG\n",
      "SWI5 ENST00000652598.1 CTG\n",
      "SWI5 ENST00000419867.7 CTG\n",
      "SWI5 ENST00000418976.3 CTG\n",
      "SWI5 ENST00000696406.1 CTG\n",
      "TEAD1 ENST00000527636.7 ATT\n",
      "TEAD1 ENST00000527575.6 ATT\n",
      "TEAD3 ENST00000402886.9 ATA\n",
      "TEAD3 ENST00000338863.13 ATA\n",
      "TEAD3 ENST00000639578.3 ATA\n",
      "TEAD4 ENST00000358409.7 TTG\n",
      "TEAD4 ENST00000359864.8 TTG\n",
      "TEAD4 ENST00000443986.3 TTG\n",
      "TRPC1 ENST00000698238.1 CTG\n",
      "TRPV6 ENST00000359396.9 ACG\n",
      "TUG1 ENST00000644773.3 CTG\n",
      "TXNRD3 ENST00000524230.9 CTG\n",
      "TXNRD3 ENST00000523403.3 CTG\n",
      "UFSP1 ENST00000672365.3 CTG\n",
      "USP19 ENST00000698200.1 AAG\n",
      "VANGL2 ENST00000696602.1 ATA\n",
      "VEGFA ENST00000672860.3 CTG\n",
      "VEGFA ENST00000372067.8 CTG\n",
      "VEGFA ENST00000425836.8 CTG\n",
      "WDR26 ENST00000414423.9 ACG\n",
      "WDR26 ENST00000678555.1 ACG\n",
      "WDR26 ENST00000678307.1 ACG\n",
      "WT1 ENST00000452863.10 CTG\n",
      "WT1 ENST00000379077.9 CTG\n",
      "WT1 ENST00000332351.9 CTG\n",
      "WT1 ENST00000448076.9 CTG\n"
     ]
    }
   ],
   "source": [
    "tmp = metadata_pc_g41[(metadata_pc_g41['utr5_end'].notna())]\n",
    "tmp2 = tmp[tmp['utr5_end'] > 0]\n",
    "\n",
    "tmp2['tr_len'] = tmp2['transcript_seq'].str.len()\n",
    "tmp3 = tmp2[tmp2['tr_len'] > tmp2['cds_stop_pos']]\n",
    "\n",
    "d = dict(tuple(tmp3[['tr_id', 'gene', 'cds_start_codon', 'utr5_end']].groupby(by='gene')))\n",
    "\n",
    "for gene, df in d.items():\n",
    "    starts = set()\n",
    "    for el in df.to_numpy():\n",
    "        tr_id = el[0]\n",
    "        cds_start_codon = el[2]\n",
    "        if cds_start_codon in ['CTG', 'TTG', 'GTG', 'ACG', 'AGG', 'AAG', 'ATT', 'ATA', 'ATC']:\n",
    "            print (gene, tr_id, cds_start_codon)\n",
    "        starts.add(cds_start_codon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ARAP1 ENST00000455638.6 ATA will be removed\n",
    "\n",
    "BAG1 ENST00000634734.3 CTG\n",
    "\n",
    "C16orf82 ENST00000505035.3 CTG - testis-specific, has goos MS spectra\n",
    "\n",
    "CACNG8 ENST00000270458.4 CTG\n",
    "\n",
    "CCDC8 ENST00000697726.1 CTG\n",
    "\n",
    "DDX17 ENST00000403230.3 CTG\n",
    "DDX17 ENST00000396821.8 CTG\n",
    "\n",
    "EBF1 ENST00000622875.4 AGG will be removed\n",
    "\n",
    "EFCAB14 ENST00000672422.2 CTG it's extension according to ribo-seq and aln but not top ranked \n",
    "\n",
    "EIF4G2 ENST00000526148.6 GTG\n",
    "EIF4G2 ENST00000525681.6 GTG\n",
    "EIF4G2 ENST00000339995.11 GTG\n",
    "EIF4G2 ENST00000396525.7 GTG\n",
    "EIF4G2 ENST00000531647.6 GTG\n",
    "EIF4G2 ENST00000528562.6 GTG\n",
    "\n",
    "FGF2 ENST00000264498.8 CTG\n",
    "\n",
    "FNDC5 ENST00000373471.9 ATA no ribo-seq for ATA. In human it's ATA, in other species its ATG   \n",
    "FNDC5 ENST00000649537.2 ATA\n",
    "\n",
    "GTF3A ENST00000381140.10 CTG\n",
    "GTF3A ENST00000435781.5 CTG\n",
    "GTF3A ENST00000419181.5 CTG\n",
    "GTF3A ENST00000438571.5 CTG\n",
    "\n",
    "HCK ENST00000486475.6 CTG\n",
    "HCK ENST00000262651.4 CTG\n",
    "HCK ENST00000375852.5 CTG\n",
    "\n",
    "HMHB1 ENST00000289448.4 CTG it's kept in MANE, it's aberrant gene \n",
    "\n",
    "IRAG1 ENST00000527509.7 CTG it has sinkORFs, dAUG is far away but no ribo-seq support\n",
    "\n",
    "KCNIP1 ENST00000390656.8 \n",
    "\n",
    "KCTD11 ENST00000333751.8 ATT, ATT in aln looks ok though upstream GUG in ribo-seq is better \n",
    " \n",
    "MDGA2 ENST00000426342.7 TTG esnembl only can be removed \n",
    "\n",
    "MORN2 ENST00000644631.4 CTG ribo-seq seems okay \n",
    "\n",
    "MRTFA ENST00000651595.2 GTG looks ok according to ribo-seq. It was not in RiboSET since there was ovlp with coding exons \n",
    "MRTFA ENST00000355630.10 GTG\n",
    "MRTFA ENST00000402042.7 GTG\n",
    "\n",
    "MYC ENST00000524013.2 CTG\n",
    "MYC ENST00000621592.8 CTG\n",
    "\n",
    "NELFB ENST00000343053.6 CTG\n",
    "\n",
    "NFKBID ENST00000641389.2 CTG\n",
    "\n",
    "NPTXR ENST00000333039.4 CTG\n",
    "\n",
    "NPW ENST00000566435.4 CTG\n",
    "\n",
    "NR1I2 ENST00000466380.6 CTG\n",
    "NR1I2 ENST00000393716.8 CTG\n",
    "\n",
    "OAZ3 ENST00000479764.7 CTG\n",
    "OAZ3 ENST00000400999.7 CTG\n",
    "\n",
    "OR8G3P ENST00000641572.1 GTG CHECK in OTTER, no ribo-seq support, only GTG isoform \n",
    "OR8G3P ENST00000641848.1 GTG\n",
    "\n",
    "PANK2 ENST00000643504.2 CTG a lot of isoforms with atg, ctg and alt 5'ends \n",
    "PANK2 ENST00000610179.7 CTG\n",
    "PANK2 ENST00000336066.8 CTG\n",
    "\n",
    "PEG10 ENST00000615790.5 CTG\n",
    "PEG10 ENST00000612941.2 CTG\n",
    "\n",
    "PIGX ENST00000392391.9 CTG\n",
    "\n",
    "POU5F1 ENST00000606567.6 CTG\n",
    "\n",
    "PRPS1L1 ENST00000506618.5 ACG\n",
    "\n",
    "PRR16 ENST00000446965.2 GTG\n",
    "\n",
    "PTEN ENST00000693560.1 CTG\n",
    "\n",
    "R3HCC1 ENST00000522012.6 CTG\n",
    "R3HCC1 ENST00000625275.3 CTG\n",
    "R3HCC1 ENST00000265806.12 CTG\n",
    "\n",
    "RNF187 ENST00000305943.9 CTG\n",
    "\n",
    "RPTOR ENST00000697423.1 CTG\n",
    "\n",
    "SFPQ ENST00000696553.1 GTG\n",
    "\n",
    "SLC22A17 ENST00000397260.7 CTG\n",
    "SLC22A17 ENST00000397267.6 CTG\n",
    "SLC22A17 ENST00000474057.6 CTG\n",
    "SLC22A17 ENST00000354772.10 CTG\n",
    "SLC22A17 ENST00000636431.2 CTG\n",
    "\n",
    "SLC6A1 ENST00000698198.1 CTG, there is dATG. no ribo-seq. why? \n",
    "\n",
    "SORCS1 ENST00000612154.4 ATA   ATA right in front of ATG. there is upstream ATG. no ribo-seq. removed? \n",
    "SORCS1 ENST00000622431.4 ATA\n",
    "\n",
    "SP3 ENST00000418194.7 ATA\n",
    "SP3 ENST00000652005.2 ATA\n",
    "\n",
    "STIM2 ENST00000467011.6 TTG\n",
    "STIM2 ENST00000465503.6 TTG\n",
    "STIM2 ENST00000467087.7 TTG\n",
    "\n",
    "SWI5 ENST00000608796.6 CTG\n",
    "SWI5 ENST00000652598.1 CTG\n",
    "SWI5 ENST00000419867.7 CTG\n",
    "SWI5 ENST00000418976.3 CTG\n",
    "SWI5 ENST00000696406.1 CTG\n",
    "\n",
    "TEAD1 ENST00000527636.7 ATT\n",
    "TEAD1 ENST00000527575.6 ATT\n",
    "\n",
    "TEAD3 ENST00000402886.9 ATA\n",
    "TEAD3 ENST00000338863.13 ATA\n",
    "TEAD3 ENST00000639578.3 ATA\n",
    "\n",
    "TEAD4 ENST00000358409.7 TTG\n",
    "TEAD4 ENST00000359864.8 TTG\n",
    "TEAD4 ENST00000443986.3 TTG\n",
    "\n",
    "TRPC1 ENST00000698238.1 CTG\n",
    "\n",
    "TRPV6 ENST00000359396.9 ACG\n",
    "\n",
    "TUG1 ENST00000644773.3 CTG   ??? \n",
    "\n",
    "TXNRD3 ENST00000524230.9 CTG\n",
    "TXNRD3 ENST00000523403.3 CTG\n",
    "\n",
    "UFSP1 ENST00000672365.3 CTG\n",
    "\n",
    "USP19 ENST00000698200.1 AAG ???\n",
    "\n",
    "VANGL2 ENST00000696602.1 ATA\n",
    "\n",
    "VEGFA ENST00000672860.3 CTG\n",
    "VEGFA ENST00000372067.8 CTG\n",
    "VEGFA ENST00000425836.8 CTG\n",
    "\n",
    "WDR26 ENST00000414423.9 ACG\n",
    "WDR26 ENST00000678555.1 ACG\n",
    "WDR26 ENST00000678307.1 ACG\n",
    "\n",
    "WT1 ENST00000452863.10 CTG\n",
    "WT1 ENST00000379077.9 CTG\n",
    "WT1 ENST00000332351.9 CTG\n",
    "WT1 ENST00000448076.9 CTG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
