{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import collections\n",
    "from collections import OrderedDict, Counter, defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from bokeh.plotting import figure, output_file, show \n",
    "import matplotlib.patches as patches    \n",
    "from matplotlib import colors\n",
    "\n",
    "\n",
    "import glob\n",
    "\n",
    "import subprocess\n",
    "from subprocess import call\n",
    "\n",
    "import re\n",
    "\n",
    "import pickle\n",
    "\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "import Bio.Data.CodonTable\n",
    "\n",
    "from scipy import stats\n",
    "\n",
    "\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "import Bio.Data.CodonTable\n",
    "\n",
    "import joypy\n",
    "\n",
    "import pyBigWig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tr_id</th>\n",
       "      <th>N_term_start</th>\n",
       "      <th>N_term_end</th>\n",
       "      <th>len</th>\n",
       "      <th>N_term_start1</th>\n",
       "      <th>N_term_end1</th>\n",
       "      <th>len_codons</th>\n",
       "      <th>gene</th>\n",
       "      <th>50len_flag</th>\n",
       "      <th>global_coo_50_and_less</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_ext</th>\n",
       "      <th>Start_ext</th>\n",
       "      <th>Stop_ext</th>\n",
       "      <th>global_rank_ext</th>\n",
       "      <th>Start_codon_ext</th>\n",
       "      <th>Proteomics_count_ext</th>\n",
       "      <th>Coverage_value_ext</th>\n",
       "      <th>CDS_ratio</th>\n",
       "      <th>Riboseq_Summary</th>\n",
       "      <th>Nstart_Nend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000624431.1</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>42</td>\n",
       "      <td>274</td>\n",
       "      <td>316</td>\n",
       "      <td>14.0</td>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>274</td>\n",
       "      <td>chr1:182667-182708</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>274-316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000624735.1</td>\n",
       "      <td>0</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "      <td>78.0</td>\n",
       "      <td>FO538757.1</td>\n",
       "      <td>86</td>\n",
       "      <td>chr1:200087-200236</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2-236</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               tr_id  N_term_start  N_term_end  len  N_term_start1  \\\n",
       "0  ENST00000624431.1           274         316   42            274   \n",
       "1  ENST00000624735.1             0         236  236              2   \n",
       "\n",
       "   N_term_end1  len_codons        gene  50len_flag global_coo_50_and_less  \\\n",
       "0          316        14.0  FO538757.2         274     chr1:182667-182708   \n",
       "1          236        78.0  FO538757.1          86     chr1:200087-200236   \n",
       "\n",
       "   ... gene_ext Start_ext Stop_ext global_rank_ext  Start_codon_ext  \\\n",
       "0  ...      NaN       NaN      NaN             NaN              NaN   \n",
       "1  ...      NaN       NaN      NaN             NaN              NaN   \n",
       "\n",
       "   Proteomics_count_ext  Coverage_value_ext  CDS_ratio  Riboseq_Summary  \\\n",
       "0                   NaN                 NaN        NaN              NaN   \n",
       "1                   NaN                 NaN        NaN              NaN   \n",
       "\n",
       "   Nstart_Nend  \n",
       "0      274-316  \n",
       "1        2-236  \n",
       "\n",
       "[2 rows x 42 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/home/alla/make_PhD_great_again/NTE/VERSION_since_May2021/'\n",
    "\n",
    "# gencode 25\n",
    "metadata_pc_g25 = pd.read_csv(path+'tmp_res/metadata_pc_g25.txt', sep='\\t')\n",
    "metadata_pc_g25['tr_id1'] = [x.split('.')[0] for x in metadata_pc_g25['tr_id'].tolist()] \n",
    "\n",
    "# gencode 41\n",
    "metadata_pc_g41 = pd.read_csv('tmp_res/metadata_pc_g41.txt', sep='\\t')\n",
    "metadata_pc_g41['tr_id1'] = [x.split('.')[0] for x in metadata_pc_g41['tr_id'].tolist()] \n",
    "metadata_pc_g41[0:2]\n",
    "\n",
    "\n",
    "# open metadata with scores and metrics\n",
    "meta = pd.read_csv('/home/alla/make_PhD_great_again/NTE/VERSION_since_May2021/tmp_res/METADATA_CURRENT.txt', sep='\\t')\n",
    "meta[0:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# nonAUG TIS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TIS Sequence</th>\n",
       "      <th>TIS Efficiency</th>\n",
       "      <th>scaled_efficiency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5977</th>\n",
       "      <td>CGCCCUGG</td>\n",
       "      <td>50.4</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6041</th>\n",
       "      <td>CGCGCUGG</td>\n",
       "      <td>50.4</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5913</th>\n",
       "      <td>CGCACUGG</td>\n",
       "      <td>50.4</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6001</th>\n",
       "      <td>CAUCCUGG</td>\n",
       "      <td>47.6</td>\n",
       "      <td>0.944444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     TIS Sequence  TIS Efficiency  scaled_efficiency\n",
       "5977     CGCCCUGG            50.4           1.000000\n",
       "6041     CGCGCUGG            50.4           1.000000\n",
       "5913     CGCACUGG            50.4           1.000000\n",
       "6001     CAUCCUGG            47.6           0.944444"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TIS_efficiency_df = pd.read_csv('/home/alla/make_PhD_great_again/NTE/VERSION_since_May2021/data/TIS_nonAUG_efficiency.txt', sep='\\t')\n",
    "\n",
    "TIS_efficiency_df['scaled_efficiency'] = TIS_efficiency_df['TIS Efficiency'] / TIS_efficiency_df['TIS Efficiency'].max()\n",
    "\n",
    "TIS_efficiency_dict = dict(zip(TIS_efficiency_df['TIS Sequence'], TIS_efficiency_df['TIS Efficiency']))\n",
    "\n",
    "TIS_efficiency_dict_scaled = dict(zip(TIS_efficiency_df['TIS Sequence'], TIS_efficiency_df['scaled_efficiency']))\n",
    "\n",
    "TIS_efficiency_df.sort_values(by='TIS Efficiency', ascending=False).head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# AUG TIS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "150 12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>efficiency</th>\n",
       "      <th>lower.bound</th>\n",
       "      <th>upper.bound</th>\n",
       "      <th>scaled_efficiency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UUCAUCAUGCA</td>\n",
       "      <td>150</td>\n",
       "      <td>139</td>\n",
       "      <td>161</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCGAGUAUGGC</td>\n",
       "      <td>149</td>\n",
       "      <td>139</td>\n",
       "      <td>161</td>\n",
       "      <td>0.993333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACGAGAAUGGC</td>\n",
       "      <td>148</td>\n",
       "      <td>136</td>\n",
       "      <td>160</td>\n",
       "      <td>0.986667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AUCAUCAUGCA</td>\n",
       "      <td>147</td>\n",
       "      <td>136</td>\n",
       "      <td>158</td>\n",
       "      <td>0.980000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACGAGUAUGGC</td>\n",
       "      <td>145</td>\n",
       "      <td>134</td>\n",
       "      <td>157</td>\n",
       "      <td>0.966667</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sequence  efficiency  lower.bound  upper.bound  scaled_efficiency\n",
       "0  UUCAUCAUGCA         150          139          161           1.000000\n",
       "1  GCGAGUAUGGC         149          139          161           0.993333\n",
       "2  ACGAGAAUGGC         148          136          160           0.986667\n",
       "3  AUCAUCAUGCA         147          136          158           0.980000\n",
       "4  ACGAGUAUGGC         145          134          157           0.966667"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TIS_AUG_eff = pd.read_csv('/home/alla/make_PhD_great_again/NTE/VERSION_since_May2021/data/TIS_AUG_efficiency.txt', sep='\\t')\n",
    "\n",
    "TIS_AUG_eff['scaled_efficiency'] = TIS_AUG_eff['efficiency'] / TIS_AUG_eff['efficiency'].max()\n",
    "\n",
    "print (TIS_AUG_eff.efficiency.max(), TIS_AUG_eff.efficiency.min())\n",
    "\n",
    "\n",
    "TIS_AUG_eff_dict = dict(zip(TIS_AUG_eff['sequence'], TIS_AUG_eff['efficiency']))\n",
    "\n",
    "\n",
    "TIS_AUG_eff_dict_scaled = dict(zip(TIS_AUG_eff['sequence'], TIS_AUG_eff['scaled_efficiency']))\n",
    "                           \n",
    "TIS_AUG_eff[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "li = []\n",
    "\n",
    "gene = 'CCDC8'\n",
    "\n",
    "tr_id = 'ENST00000307522'\n",
    "\n",
    "pre_st = '565_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "gene = 'EIF4G2'\n",
    "\n",
    "tr_id = 'ENST00000339995'\n",
    "\n",
    "pre_st = '493_GTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "gene = 'BAG1'\n",
    "\n",
    "tr_id = 'ENST00000472232'\n",
    "\n",
    "pre_st = '88_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "gene = 'GTF3A'\n",
    "\n",
    "tr_id = 'ENST00000381140'\n",
    "\n",
    "pre_st = '195_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "gene = 'DDX17'\n",
    "\n",
    "tr_id = 'ENST00000396821'\n",
    "\n",
    "pre_st = '101_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "gene = 'RNF187'\n",
    "\n",
    "tr_id = 'ENST00000305943'\n",
    "\n",
    "pre_st = '429_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "\n",
    "gene = 'NELFB'\n",
    "\n",
    "tr_id = 'ENST00000343053'\n",
    "\n",
    "pre_st = '194_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "\n",
    "gene = 'NPTXR'\n",
    "\n",
    "tr_id = 'ENST00000333039'\n",
    "\n",
    "pre_st = '125_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "\n",
    "gene = 'NPW'\n",
    "\n",
    "tr_id = 'ENST00000329610'\n",
    "\n",
    "pre_st = '383_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "\n",
    "gene = 'PIGX'\n",
    "\n",
    "tr_id = 'ENST00000392391'\n",
    "\n",
    "pre_st = '176_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "gene = 'PRPS1L1'\n",
    "\n",
    "tr_id = 'ENST00000506618'\n",
    "\n",
    "pre_st = '82_ACG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "\n",
    "#gene = 'PRR16'\n",
    "\n",
    "#tr_id = 'ENST00000446965'\n",
    "\n",
    "#pre_st = '82_GTG'\n",
    "\n",
    "#li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "gene = 'SLC22A17'\n",
    "\n",
    "tr_id = 'ENST00000397260'\n",
    "\n",
    "pre_st = '120_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "\n",
    "gene = 'SP3'\n",
    "\n",
    "tr_id = 'ENST00000418194'\n",
    "\n",
    "pre_st = '385_ATA'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "gene = 'SWI5'\n",
    "\n",
    "tr_id = 'ENST00000608796'\n",
    "\n",
    "pre_st = '120_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "\n",
    "gene = 'TXNRD3'\n",
    "\n",
    "tr_id = 'ENST00000524230'\n",
    "\n",
    "pre_st = '166_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "\n",
    "gene = 'R3HCC1'\n",
    "\n",
    "tr_id = 'ENST00000265806'\n",
    "\n",
    "pre_st = '269_CTG'\n",
    "\n",
    "li.append([gene, tr_id, int(pre_st.split('_')[0]), pre_st.split('_')[1]])\n",
    "\n",
    "df = pd.DataFrame(li, columns = ['gene', 'tr_id1', 'start', 'codon'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_pc_g25['tr_id1'] = [x.split('.')[0] for x in metadata_pc_g25['tr_id'].tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tr_id1</th>\n",
       "      <th>start</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDC8</td>\n",
       "      <td>ENST00000307522</td>\n",
       "      <td>565</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>ENST00000339995</td>\n",
       "      <td>493</td>\n",
       "      <td>GTG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene           tr_id1  start codon\n",
       "0   CCDC8  ENST00000307522    565   CTG\n",
       "1  EIF4G2  ENST00000339995    493   GTG"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[0:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDC8</td>\n",
       "      <td>ENST00000307522</td>\n",
       "      <td>565</td>\n",
       "      <td>CTG</td>\n",
       "      <td>CACCCTGGAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>ENST00000339995</td>\n",
       "      <td>493</td>\n",
       "      <td>GTG</td>\n",
       "      <td>CAAAGTGGAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BAG1</td>\n",
       "      <td>ENST00000472232</td>\n",
       "      <td>88</td>\n",
       "      <td>CTG</td>\n",
       "      <td>GGGCCTGGCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTF3A</td>\n",
       "      <td>ENST00000381140</td>\n",
       "      <td>195</td>\n",
       "      <td>CTG</td>\n",
       "      <td>CGCCCTGGATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DDX17</td>\n",
       "      <td>ENST00000396821</td>\n",
       "      <td>101</td>\n",
       "      <td>CTG</td>\n",
       "      <td>AAACCTGCCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RNF187</td>\n",
       "      <td>ENST00000305943</td>\n",
       "      <td>429</td>\n",
       "      <td>CTG</td>\n",
       "      <td>AGCCCTGGCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NELFB</td>\n",
       "      <td>ENST00000343053</td>\n",
       "      <td>194</td>\n",
       "      <td>CTG</td>\n",
       "      <td>GGACCTGGCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NPTXR</td>\n",
       "      <td>ENST00000333039</td>\n",
       "      <td>125</td>\n",
       "      <td>CTG</td>\n",
       "      <td>CACGCTGAAGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NPW</td>\n",
       "      <td>ENST00000329610</td>\n",
       "      <td>383</td>\n",
       "      <td>CTG</td>\n",
       "      <td>CGCCCTGGCGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PIGX</td>\n",
       "      <td>ENST00000392391</td>\n",
       "      <td>176</td>\n",
       "      <td>CTG</td>\n",
       "      <td>CGTCCTGGCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PRPS1L1</td>\n",
       "      <td>ENST00000506618</td>\n",
       "      <td>82</td>\n",
       "      <td>ACG</td>\n",
       "      <td>CAAGACGCCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SLC22A17</td>\n",
       "      <td>ENST00000397260</td>\n",
       "      <td>120</td>\n",
       "      <td>CTG</td>\n",
       "      <td>CAAGCTGGCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SP3</td>\n",
       "      <td>ENST00000418194</td>\n",
       "      <td>385</td>\n",
       "      <td>ATA</td>\n",
       "      <td>CAAGATAGGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SWI5</td>\n",
       "      <td>ENST00000608796</td>\n",
       "      <td>120</td>\n",
       "      <td>CTG</td>\n",
       "      <td>CGCGCTGGACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TXNRD3</td>\n",
       "      <td>ENST00000524230</td>\n",
       "      <td>166</td>\n",
       "      <td>CTG</td>\n",
       "      <td>GACTCTGGAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>R3HCC1</td>\n",
       "      <td>ENST00000265806</td>\n",
       "      <td>269</td>\n",
       "      <td>CTG</td>\n",
       "      <td>CACCCTGGCCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0                1    2    3            4\n",
       "0      CCDC8  ENST00000307522  565  CTG  CACCCTGGAGA\n",
       "1     EIF4G2  ENST00000339995  493  GTG  CAAAGTGGAGA\n",
       "2       BAG1  ENST00000472232   88  CTG  GGGCCTGGCTC\n",
       "3      GTF3A  ENST00000381140  195  CTG  CGCCCTGGATC\n",
       "4      DDX17  ENST00000396821  101  CTG  AAACCTGCCCA\n",
       "5     RNF187  ENST00000305943  429  CTG  AGCCCTGGCGC\n",
       "6      NELFB  ENST00000343053  194  CTG  GGACCTGGCCG\n",
       "7      NPTXR  ENST00000333039  125  CTG  CACGCTGAAGT\n",
       "8        NPW  ENST00000329610  383  CTG  CGCCCTGGCGT\n",
       "9       PIGX  ENST00000392391  176  CTG  CGTCCTGGCGG\n",
       "10   PRPS1L1  ENST00000506618   82  ACG  CAAGACGCCGA\n",
       "11  SLC22A17  ENST00000397260  120  CTG  CAAGCTGGCTC\n",
       "12       SP3  ENST00000418194  385  ATA  CAAGATAGGGC\n",
       "13      SWI5  ENST00000608796  120  CTG  CGCGCTGGACC\n",
       "14    TXNRD3  ENST00000524230  166  CTG  GACTCTGGAGC\n",
       "15    R3HCC1  ENST00000265806  269  CTG  CACCCTGGCCC"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "li = []\n",
    "for el in df.to_numpy():\n",
    "    seq = metadata_pc_g25[metadata_pc_g25['tr_id1'] == el[1]].iloc[0].transcript_seq\n",
    "    start = el[2]\n",
    "    nonAUG_context = seq[start-5:start+6]\n",
    "    li.append([el[0], el[1], el[2], el[3], nonAUG_context])\n",
    "    \n",
    "    # nonAUG TIS score \n",
    "\n",
    "pd.DataFrame(li)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tr_id1</th>\n",
       "      <th>start</th>\n",
       "      <th>codon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCDC8</td>\n",
       "      <td>ENST00000307522</td>\n",
       "      <td>565</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>ENST00000339995</td>\n",
       "      <td>493</td>\n",
       "      <td>GTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BAG1</td>\n",
       "      <td>ENST00000472232</td>\n",
       "      <td>88</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTF3A</td>\n",
       "      <td>ENST00000381140</td>\n",
       "      <td>195</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DDX17</td>\n",
       "      <td>ENST00000396821</td>\n",
       "      <td>101</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RNF187</td>\n",
       "      <td>ENST00000305943</td>\n",
       "      <td>429</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NELFB</td>\n",
       "      <td>ENST00000343053</td>\n",
       "      <td>194</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NPTXR</td>\n",
       "      <td>ENST00000333039</td>\n",
       "      <td>125</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NPW</td>\n",
       "      <td>ENST00000329610</td>\n",
       "      <td>383</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PIGX</td>\n",
       "      <td>ENST00000392391</td>\n",
       "      <td>176</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PRPS1L1</td>\n",
       "      <td>ENST00000506618</td>\n",
       "      <td>82</td>\n",
       "      <td>ACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SLC22A17</td>\n",
       "      <td>ENST00000397260</td>\n",
       "      <td>120</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SP3</td>\n",
       "      <td>ENST00000418194</td>\n",
       "      <td>385</td>\n",
       "      <td>ATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SWI5</td>\n",
       "      <td>ENST00000608796</td>\n",
       "      <td>120</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TXNRD3</td>\n",
       "      <td>ENST00000524230</td>\n",
       "      <td>166</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>R3HCC1</td>\n",
       "      <td>ENST00000265806</td>\n",
       "      <td>269</td>\n",
       "      <td>CTG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene           tr_id1  start codon\n",
       "0      CCDC8  ENST00000307522    565   CTG\n",
       "1     EIF4G2  ENST00000339995    493   GTG\n",
       "2       BAG1  ENST00000472232     88   CTG\n",
       "3      GTF3A  ENST00000381140    195   CTG\n",
       "4      DDX17  ENST00000396821    101   CTG\n",
       "5     RNF187  ENST00000305943    429   CTG\n",
       "6      NELFB  ENST00000343053    194   CTG\n",
       "7      NPTXR  ENST00000333039    125   CTG\n",
       "8        NPW  ENST00000329610    383   CTG\n",
       "9       PIGX  ENST00000392391    176   CTG\n",
       "10   PRPS1L1  ENST00000506618     82   ACG\n",
       "11  SLC22A17  ENST00000397260    120   CTG\n",
       "12       SP3  ENST00000418194    385   ATA\n",
       "13      SWI5  ENST00000608796    120   CTG\n",
       "14    TXNRD3  ENST00000524230    166   CTG\n",
       "15    R3HCC1  ENST00000265806    269   CTG"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TIS score of sink ORFs and nonAUG starts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "could be actually like a two-sided barplot with showing top-score as dashed line for both types of codons "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Any polyAla downstream of nonAUG starts? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Any quadruplexes? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
